File: code/go/ch19/src/luago/compiler/parser/optimizer.go
package parser

import "math"
import "luago/number"
import . "luago/compiler/ast"
import . "luago/compiler/lexer"

func optimizeLogicalOr(exp *BinopExp) Exp {
	if isTrue(exp.Exp1) {
		return exp.Exp1 // true or x => true
	}
	if isFalse(exp.Exp1) && !isVarargOrFuncCall(exp.Exp2) {
		return exp.Exp2 // false or x => x
	}
	return exp
}

func optimizeLogicalAnd(exp *BinopExp) Exp {
	if isFalse(exp.Exp1) {
		return exp.Exp1 // false and x => false
	}
	if isTrue(exp.Exp1) && !isVarargOrFuncCall(exp.Exp2) {
		return exp.Exp2 // true and x => x
	}
	return exp
}

func optimizeBitwiseBinaryOp(exp *BinopExp) Exp {
	if i, ok := castToInt(exp.Exp1); ok {
		if j, ok := castToInt(exp.Exp2); ok {
			switch exp.Op {
			case TOKEN_OP_BAND:
				return &IntegerExp{exp.Line, i & j}
			case TOKEN_OP_BOR:
				return &IntegerExp{exp.Line, i | j}
			case TOKEN_OP_BXOR:
				return &IntegerExp{exp.Line, i ^ j}
			case TOKEN_OP_SHL:
				return &IntegerExp{exp.Line, number.ShiftLeft(i, j)}
			case TOKEN_OP_SHR:
				return &IntegerExp{exp.Line, number.ShiftRight(i, j)}
			}
		}
	}
	return exp
}

func optimizeArithBinaryOp(exp *BinopExp) Exp {
	if x, ok := exp.Exp1.(*IntegerExp); ok {
		if y, ok := exp.Exp2.(*IntegerExp); ok {
			switch exp.Op {
			case TOKEN_OP_ADD:
				return &IntegerExp{exp.Line, x.Val + y.Val}
			case TOKEN_OP_SUB:
				return &IntegerExp{exp.Line, x.Val - y.Val}
			case TOKEN_OP_MUL:
				return &IntegerExp{exp.Line, x.Val * y.Val}
			case TOKEN_OP_IDIV:
				if y.Val != 0 {
					return &IntegerExp{exp.Line, number.IFloorDiv(x.Val, y.Val)}
				}
			case TOKEN_OP_MOD:
				if y.Val != 0 {
					return &IntegerExp{exp.Line, number.IMod(x.Val, y.Val)}
				}
			}
		}
	}
	if f, ok := castToFloat(exp.Exp1); ok {
		if g, ok := castToFloat(exp.Exp2); ok {
			switch exp.Op {
			case TOKEN_OP_ADD:
				return &FloatExp{exp.Line, f + g}
			case TOKEN_OP_SUB:
				return &FloatExp{exp.Line, f - g}
			case TOKEN_OP_MUL:
				return &FloatExp{exp.Line, f * g}
			case TOKEN_OP_DIV:
				if g != 0 {
					return &FloatExp{exp.Line, f / g}
				}
			case TOKEN_OP_IDIV:
				if g != 0 {
					return &FloatExp{exp.Line, number.FFloorDiv(f, g)}
				}
			case TOKEN_OP_MOD:
				if g != 0 {
					return &FloatExp{exp.Line, number.FMod(f, g)}
				}
			case TOKEN_OP_POW:
				return &FloatExp{exp.Line, math.Pow(f, g)}
			}
		}
	}
	return exp
}

func optimizePow(exp Exp) Exp {
	if binop, ok := exp.(*BinopExp); ok {
		if binop.Op == TOKEN_OP_POW {
			binop.Exp2 = optimizePow(binop.Exp2)
		}
		return optimizeArithBinaryOp(binop)
	}
	return exp
}

func optimizeUnaryOp(exp *UnopExp) Exp {
	switch exp.Op {
	case TOKEN_OP_UNM:
		return optimizeUnm(exp)
	case TOKEN_OP_NOT:
		return optimizeNot(exp)
	case TOKEN_OP_BNOT:
		return optimizeBnot(exp)
	default:
		return exp
	}
}

func optimizeUnm(exp *UnopExp) Exp {
	switch x := exp.Exp.(type) { // number?
	case *IntegerExp:
		x.Val = -x.Val
		return x
	case *FloatExp:
		if x.Val != 0 {
			x.Val = -x.Val
			return x
		}
	}
	return exp
}

func optimizeNot(exp *UnopExp) Exp {
	switch exp.Exp.(type) {
	case *NilExp, *FalseExp: // false
		return &TrueExp{exp.Line}
	case *TrueExp, *IntegerExp, *FloatExp, *StringExp: // true
		return &FalseExp{exp.Line}
	default:
		return exp
	}
}

func optimizeBnot(exp *UnopExp) Exp {
	switch x := exp.Exp.(type) { // number?
	case *IntegerExp:
		x.Val = ^x.Val
		return x
	case *FloatExp:
		if i, ok := number.FloatToInteger(x.Val); ok {
			return &IntegerExp{x.Line, ^i}
		}
	}
	return exp
}

func isFalse(exp Exp) bool {
	switch exp.(type) {
	case *FalseExp, *NilExp:
		return true
	default:
		return false
	}
}

func isTrue(exp Exp) bool {
	switch exp.(type) {
	case *TrueExp, *IntegerExp, *FloatExp, *StringExp:
		return true
	default:
		return false
	}
}

// todo
func isVarargOrFuncCall(exp Exp) bool {
	switch exp.(type) {
	case *VarargExp, *FuncCallExp:
		return true
	}
	return false
}

func castToInt(exp Exp) (int64, bool) {
	switch x := exp.(type) {
	case *IntegerExp:
		return x.Val, true
	case *FloatExp:
		return number.FloatToInteger(x.Val)
	default:
		return 0, false
	}
}

func castToFloat(exp Exp) (float64, bool) {
	switch x := exp.(type) {
	case *IntegerExp:
		return float64(x.Val), true
	case *FloatExp:
		return x.Val, true
	default:
		return 0, false
	}
}

File: code/java/ch17/src/main/java/com/github/zxh0/luago/compiler/parser/Optimizer.java
package com.github.zxh0.luago.compiler.parser;

import com.github.zxh0.luago.compiler.ast.Exp;
import com.github.zxh0.luago.compiler.ast.exps.*;
import com.github.zxh0.luago.number.LuaMath;
import com.github.zxh0.luago.number.LuaNumber;

import static com.github.zxh0.luago.compiler.lexer.TokenKind.TOKEN_OP_POW;

class Optimizer {

    static Exp optimizeLogicalOr(BinopExp exp) {
        if (isTrue(exp.getExp1())) {
            return exp.getExp1(); // true or x => true
        }
        if (isFalse(exp.getExp1()) && !isVarargOrFuncCall(exp.getExp2())) {
            return exp.getExp2(); // false or x => x
        }
        return exp;
    }

    static Exp optimizeLogicalAnd(BinopExp exp) {
        if (isFalse(exp.getExp1())) {
            return exp.getExp1(); // false and x => false
        }
        if (isTrue(exp.getExp1()) && !isVarargOrFuncCall(exp.getExp2())) {
            return exp.getExp2(); // true and x => x
        }
        return exp;
    }

    static Exp optimizeBitwiseBinaryOp(BinopExp exp) {
        Long i = castToInteger(exp.getExp1());
        if (i != null) {
            Long j = castToInteger(exp.getExp2());
            if (j != null) {
                switch (exp.getOp()) {
                    case TOKEN_OP_BAND: return new IntegerExp(exp.getLine(), i & j);
                    case TOKEN_OP_BOR:  return new IntegerExp(exp.getLine(), i | j);
                    case TOKEN_OP_BXOR: return new IntegerExp(exp.getLine(), i ^ j);
                    case TOKEN_OP_SHL:  return new IntegerExp(exp.getLine(), LuaMath.shiftLeft(i, j));
                    case TOKEN_OP_SHR:  return new IntegerExp(exp.getLine(), LuaMath.shiftRight(i, j));
                }
            }
        }
        return exp;
    }

    static Exp optimizeArithBinaryOp(BinopExp exp) {
        if (exp.getExp1() instanceof IntegerExp
                && exp.getExp2() instanceof IntegerExp) {
            IntegerExp x = (IntegerExp) exp.getExp1();
            IntegerExp y = (IntegerExp) exp.getExp2();
            switch (exp.getOp()) {
                case TOKEN_OP_ADD: return new IntegerExp(exp.getLine(), x.getVal() + y.getVal());
                case TOKEN_OP_SUB: return new IntegerExp(exp.getLine(), x.getVal() - y.getVal());
                case TOKEN_OP_MUL: return new IntegerExp(exp.getLine(), x.getVal() * y.getVal());
                case TOKEN_OP_IDIV:
                    if (y.getVal() != 0) {
                        return new IntegerExp(exp.getLine(), Math.floorDiv(x.getVal(), y.getVal()));
                    }
                    break;
                case TOKEN_OP_MOD:
                    if (y.getVal() != 0) {
                        return new IntegerExp(exp.getLine(), Math.floorMod(x.getVal(), y.getVal()));
                    }
                    break;
            }
        }

        Double f = castToFloat(exp.getExp1());
        if (f != null) {
            Double g = castToFloat(exp.getExp2());
            if (g != null) {
                switch (exp.getOp()) {
                    case TOKEN_OP_ADD: return new FloatExp(exp.getLine(), f + g);
                    case TOKEN_OP_SUB: return new FloatExp(exp.getLine(), f - g);
                    case TOKEN_OP_MUL: return new FloatExp(exp.getLine(), f * g);
                    case TOKEN_OP_POW: return new FloatExp(exp.getLine(), Math.pow(f, g));
                }
                if (g != 0) {
                    switch (exp.getOp()) {
                        case TOKEN_OP_DIV:  return new FloatExp(exp.getLine(), f / g);
                        case TOKEN_OP_IDIV: return new FloatExp(exp.getLine(), LuaMath.floorDiv(f, g));
                        case TOKEN_OP_MOD:  return new FloatExp(exp.getLine(), LuaMath.floorMod(f, g));
                    }
                }
            }
        }

        return exp;
    }

    static Exp optimizePow(Exp exp) {
        if (exp instanceof BinopExp) {
            BinopExp binopExp = (BinopExp) exp;
            if (binopExp.getOp() == TOKEN_OP_POW) {
                binopExp.setExp2(optimizePow(binopExp.getExp2()));
            }
            return optimizeArithBinaryOp(binopExp);
        }
        return exp;
    }

    static Exp optimizeUnaryOp(UnopExp exp) {
        switch (exp.getOp()) {
            case TOKEN_OP_UNM:  return optimizeUnm(exp);
            case TOKEN_OP_NOT:  return optimizeNot(exp);
            case TOKEN_OP_BNOT: return optimizeBnot(exp);
            default: return exp;
        }
    }

    private static Exp optimizeUnm(UnopExp exp) {
        if (exp.getExp() instanceof IntegerExp) {
            IntegerExp iExp = (IntegerExp) exp.getExp();
            iExp.setVal(-iExp.getVal());
            return iExp;
        }
        if (exp.getExp() instanceof FloatExp) {
            FloatExp fExp = (FloatExp) exp.getExp();
            fExp.setVal(-fExp.getVal());
            return fExp;
        }
        return exp;
    }

    private static Exp optimizeNot(UnopExp exp) {
        Exp subExp = exp.getExp();
        if (subExp instanceof NilExp
                || subExp instanceof FalseExp) {
            return new TrueExp(exp.getLine());
        }
        if (subExp instanceof TrueExp
                || subExp instanceof IntegerExp
                || subExp instanceof FloatExp
                || subExp instanceof StringExp) {
            return new FalseExp(exp.getLine());
        }
        return exp;
    }

    private static Exp optimizeBnot(UnopExp exp) {
        if (exp.getExp() instanceof IntegerExp) {
            IntegerExp iExp = (IntegerExp) exp.getExp();
            iExp.setVal(~iExp.getVal());
            return iExp;
        }
        if (exp.getExp() instanceof FloatExp) {
            FloatExp fExp = (FloatExp) exp.getExp();
            double f = fExp.getVal();
            if (LuaNumber.isInteger(f)) {
                return new IntegerExp(fExp.getLine(), ~((int) f));
            }
        }
        return exp;
    }

    private static boolean isFalse(Exp exp) {
        return exp instanceof FalseExp
                || exp instanceof NilExp;
    }

    private static boolean isTrue(Exp exp) {
        return exp instanceof TrueExp
                || exp instanceof IntegerExp
                || exp instanceof FloatExp
                || exp instanceof StringExp;
    }

    private static boolean isVarargOrFuncCall(Exp exp) {
        return exp instanceof VarargExp
                || exp instanceof FuncCallExp;
    }

    private static Long castToInteger(Exp exp) {
        if (exp instanceof IntegerExp) {
            return ((IntegerExp) exp).getVal();
        }
        if (exp instanceof FloatExp) {
            double f = ((FloatExp) exp).getVal();
            return LuaNumber.isInteger(f) ? (long) f : null;
        }
        return null;
    }

    private static Double castToFloat(Exp exp) {
        if (exp instanceof IntegerExp) {
            return (double) ((IntegerExp) exp).getVal();
        }
        if (exp instanceof FloatExp) {
            return ((FloatExp) exp).getVal();
        }
        return null;
    }

}
